ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/services/ApiSubmissionService.scala

```diff
@@ -243,15 +246,32 @@ final class ApiSubmissionService private (
       case Some(result) =>
         Future.successful(result)
       case None =>
-        transactionInfo match {
-          case CommandExecutionResult(submitterInfo, transactionMeta, transaction, _) =>
-            Timed.future(
-              Metrics.submittedTransactionsTimer,
-              FutureConverters.toScala(
-                writeService.submitTransaction(submitterInfo, transactionMeta, transaction)))
+        // Submit transactions such that they arrive at the ledger sequencer exactly when record time equals ledger time.
+        // If the ledger time of the transaction is far in the future (farther than the expected latency),
+        // the submission to the WriteService is delayed.
+        val submitAt = transactionInfo.transactionMeta.ledgerEffectiveTime.toInstant
+          .minus(timeModel.avgTransactionLatency)
+        val submissionDelay = Duration.between(Instant.now, submitAt)
+        if (submissionDelay.isNegative)
+          submitTransaction(transactionInfo)
+        else {
+          Metrics.delayedSubmissionsMeter.mark()
+          val scalaDelay = scala.concurrent.duration.Duration.fromNanos(submissionDelay.toNanos)
+          Delayed.Future.by(scalaDelay)(submitTransaction(transactionInfo))
```

Comment:
Is this performant enough? `Delayed.Future` uses a singleton daemon `java.util.Timer`.

Reply:
`java.util.Timer` should be able to scale to thousands of timers without any issue.